### PR TITLE
[js/webgpu] Enable pad f16 uniform

### PR DESCRIPTION
**Pull request by axinging**

### CODE REVIEW

**Comment by axinging** [comment]
@qjia7 @xhcao @hujiajie @jzm-intel @gyagp PTAL

File: js/web/lib/wasm/jsep/backend-webgpu.ts [hunk 1]
```diff
@@ -538,8 +538,14 @@ export class WebGpuBackend {
...
         } else if (v.type === DataType.float16) {
-          // TODO: use Float16Array.
-          new Uint16Array(arrayBuffer, offset, data.length).set(data);
+          if (typeof Float16Array !== 'undefined') {
+            new Float16Array(arrayBuffer, offset, data.length).set(data);
+          } else {
+            // Fallback to Uint16Array when Float16Array polyfill is not available, unit test only.
+            // eslint-disable-next-line no-console
+            console.warn('Unit test only, please make sure the float16 data has been encoded as float 16 bits.');
+            new Uint16Array(arrayBuffer, offset, data.length).set(data);
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/backend-webgpu.ts:547]: My understanding is that we always use `Uint16Array` for float16 data. So there is no need to do this change?

File: js/web/lib/wasm/jsep/webgpu/ops/pad.ts [hunk 2]
```diff
@@ -200,7 +200,9 @@ const createPadProgramInfo = (inputs: readonly TensorView[], attributes: PadAttr
 const createPadAttributesFromInputs = (inputs: readonly TensorView[], attributes: PadAttributes): PadAttributes => {
   if (inputs.length > 1) {
     const bigInt64Pads = inputs[1].getBigInt64Array();
-    const value = (inputs.length >= 3 && inputs[2].data) ? inputs[2].getFloat32Array()[0] : 0.0;
+    const value = (inputs.length >= 3 && inputs[2].data) ?
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/webgpu/ops/pad.ts:203]: Please use uniform for `constant_value`. So if `inputs[2].dataType` is  `DataType.float16`, you can add a `getUint16Array()` and write it to a uniform buffer which is similar like the discussion in #21584.

File: js/web/lib/wasm/jsep/webgpu/ops/pad.ts [hunk 3]
```diff
@@ -154,8 +154,14 @@ const createPadProgramInfo = (inputs: readonly TensorView[], attributes: PadAttr
   const outputSize = ShapeUtil.size(outputShape);
   const programUniforms: ProgramUniform[] =
       [{type: DataType.uint32, data: outputSize}, {type: DataType.int32, data: attributes.pads}];
+
+  const isValueFromInput = inputs.length >= 3 && inputs[2].data;
   if (attributes.mode === 0) {
-    programUniforms.push({type: inputs[0].dataType, data: attributes.value});
+    programUniforms.push({
+      type: inputs[0].dataType === DataType.float16 ? (isValueFromInput ? DataType.float16 : DataType.float) :
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/webgpu/ops/pad.ts:161]: `type: isValueFromInput ? inputs[2].dataType : DataType.float`

File: js/web/lib/wasm/jsep/webgpu/ops/pad.ts [hunk 4]
```diff
@@ -169,7 +175,10 @@ const createPadProgramInfo = (inputs: readonly TensorView[], attributes: PadAttr
     const uniforms: UniformsArrayType =
         [{name: 'output_size', type: 'u32'}, {name: 'pads', type: 'i32', length: attributes.pads.length}];
     if (attributes.mode === 0) {
-      uniforms.push({name: 'constant_value', type: dataType as UniformDataElementType});
+      uniforms.push({
+        name: 'constant_value',
+        type: dataType === 'f16' ? (isValueFromInput ? 'f16' : 'f32') : dataType as UniformDataElementType
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/webgpu/ops/pad.ts:180]: Not necessary to change here?

File: js/web/lib/wasm/jsep/webgpu/ops/pad.ts (hunk 3 above)
Comment (axinging) [reply js/web/lib/wasm/jsep/webgpu/ops/pad.ts]: done, thanks!

File: js/web/lib/wasm/jsep/webgpu/ops/pad.ts (hunk 4 above)
Comment (axinging) [reply js/web/lib/wasm/jsep/webgpu/ops/pad.ts]: This workarounds uniform checks

**Review by axinging — COMMENTED** [review]
@qjia7, PTAL

**Review by qjia7 — APPROVED** [review]
LGTM, thanks.

@guschmue @fs-eire @satyajandhyala Please take a look, thanks.

**Review by fs-eire — DISMISSED** [review]
I did merge the code to main branch so technically I cannot approve code of myself. Need another code owner to approve.

**Comment by guschmue** [comment]
lint unhappy

**Comment by fs-eire** [comment]
will take a look

> lint unhappy

**Review by fs-eire — APPROVED** [review]

**Review by guschmue — APPROVED** [review]

**Pull request merged**